File: k8s-dws-controller/src/queue_manager.py
"""Priority-based queue manager with per-queue concurrency limits.

Jobs are ordered by priority (descending) then by name (ascending, as a
stable FIFO proxy).  Each named queue enforces an independent concurrency
cap so that one busy queue cannot starve another.
"""

from __future__ import annotations

import logging
import threading
from dataclasses import dataclass, field

import kubernetes.client as k8s_client
from kubernetes.client.rest import ApiException

from .models import API_GROUP, API_VERSION, PLURAL, Phase

logger = logging.getLogger(__name__)


@dataclass(order=False)
class _QueueEntry:
    name: str
    namespace: str
    priority: int


class QueueManager:
    """In-memory priority queue backed by the cluster for concurrency checks."""

    def __init__(self, max_concurrent_per_queue: int = 5) -> None:
        self._lock = threading.Lock()
        self._queues: dict[str, list[_QueueEntry]] = {}
        self.max_concurrent_per_queue = max_concurrent_per_queue

    # ── helpers ──────────────────────────────────────────────────────────

    @staticmethod
    def _sort_key(entry: _QueueEntry) -> tuple[int, str]:
        """Higher priority first, then alphabetical name (FIFO proxy)."""
        return (-entry.priority, entry.name)

    def _sort_queue(self, queue_name: str) -> None:
        self._queues.setdefault(queue_name, []).sort(key=self._sort_key)

    def _find(self, queue_name: str, name: str, namespace: str) -> int | None:
        """Return index in the named queue, or None."""
        for i, e in enumerate(self._queues.get(queue_name, [])):
            if e.name == name and e.namespace == namespace:
                return i
        return None

    # ── public API ───────────────────────────────────────────────────────

    def enqueue(self, name: str, namespace: str, priority: int, queue_name: str) -> int:
        """Add a job to its queue.  Returns 1-based position."""
        queue_name = queue_name or "default"
        with self._lock:
            idx = self._find(queue_name, name, namespace)
            if idx is not None:
                return idx + 1

            self._queues.setdefault(queue_name, []).append(
                _QueueEntry(name=name, namespace=namespace, priority=priority)
            )
            self._sort_queue(queue_name)

            pos = self._find(queue_name, name, namespace)
            return (pos + 1) if pos is not None else len(self._queues[queue_name])

    def dequeue(self, name: str, namespace: str, queue_name: str) -> None:
        """Remove a job from its queue."""
        queue_name = queue_name or "default"
        with self._lock:
            idx = self._find(queue_name, name, namespace)
            if idx is not None:
                self._queues[queue_name].pop(idx)

    def get_position(self, name: str, namespace: str, queue_name: str) -> int:
        """Return 1-based queue position, or 0 if not queued."""
        queue_name = queue_name or "default"
        with self._lock:
            idx = self._find(queue_name, name, namespace)
            return (idx + 1) if idx is not None else 0

    def can_admit(self, name: str, namespace: str, queue_name: str) -> bool:
        """Check whether a job may be admitted from its queue.

        Admission requires:
          1. The job is at position 1 (front of queue).
          2. Active (Provisioning | Running) jobs in this queue are below
             the concurrency cap.
        """
        queue_name = queue_name or "default"

        with self._lock:
            idx = self._find(queue_name, name, namespace)
            if idx is None or idx != 0:
                return False

        # Count active jobs in this queue via the K8s API.
        active = self._count_active_in_queue(queue_name)
        if active >= self.max_concurrent_per_queue:
            logger.info(
                "Queue %r at concurrency limit (%d/%d)",
                queue_name,
                active,
                self.max_concurrent_per_queue,
            )
            return False

        return True

    # ── cluster interaction ──────────────────────────────────────────────

    def _count_active_in_queue(self, queue_name: str) -> int:
        """Count Provisioning/Running DWSJobs in *queue_name*."""
        api = k8s_client.CustomObjectsApi()
        try:
            resp = api.list_cluster_custom_object(
                group=API_GROUP,
                version=API_VERSION,
                plural=PLURAL,
            )
        except ApiException as exc:
            logger.error("Failed to list DWSJobs: %s", exc)
            return 0

        active = 0
        for item in resp.get("items", []):
            spec = item.get("spec", {})
            status = item.get("status", {})
            item_queue = spec.get("queueName") or "default"
            if item_queue != queue_name:
                continue
            if status.get("phase") in (Phase.PROVISIONING, Phase.RUNNING):
                active += 1
        return active

    def rebuild_from_cluster(self) -> None:
        """Rebuild in-memory state from the cluster (called on startup)."""
        api = k8s_client.CustomObjectsApi()
        try:
            resp = api.list_cluster_custom_object(
                group=API_GROUP,
                version=API_VERSION,
                plural=PLURAL,
            )
        except ApiException as exc:
            logger.error("Failed to list DWSJobs for queue rebuild: %s", exc)
            return

        with self._lock:
            self._queues.clear()

            for item in resp.get("items", []):
                spec = item.get("spec", {})
                status = item.get("status", {})
                phase = status.get("phase", Phase.PENDING)

                if phase not in (Phase.PENDING, Phase.QUEUED):
                    continue

                queue_name = spec.get("queueName") or "default"
                entry = _QueueEntry(
                    name=item["metadata"]["name"],
                    namespace=item["metadata"]["namespace"],
                    priority=spec.get("priority", 0),
                )
                self._queues.setdefault(queue_name, []).append(entry)

            for qn in self._queues:
                self._sort_queue(qn)
                logger.info("Rebuilt queue %r  (%d entries)", qn, len(self._queues[qn]))

File: k8s-dws-controller/internal/controller/queue_manager.go
package controller

import (
	"context"
	"sort"
	"sync"

	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/log"

	dwsv1alpha1 "github.com/chartotu19/humdrum/k8s-dws-controller/api/v1alpha1"
)

// QueueManager manages job queues with priority-based ordering and
// per-queue concurrency limits to provide fair scheduling across queues.
type QueueManager struct {
	client client.Client
	mu     sync.RWMutex

	// maxConcurrentPerQueue limits how many jobs can be in Provisioning or Running
	// state per queue at the same time.
	maxConcurrentPerQueue int

	// queues tracks queued DWSJob names per queue, ordered by priority (descending)
	// then by creation timestamp (ascending, FIFO within same priority).
	queues map[string][]queueEntry
}

type queueEntry struct {
	name      string
	namespace string
	priority  int32
}

// NewQueueManager creates a new QueueManager.
func NewQueueManager(c client.Client, maxConcurrentPerQueue int) *QueueManager {
	if maxConcurrentPerQueue <= 0 {
		maxConcurrentPerQueue = 5
	}
	return &QueueManager{
		client:                c,
		maxConcurrentPerQueue: maxConcurrentPerQueue,
		queues:                make(map[string][]queueEntry),
	}
}

// Enqueue adds a job to its queue. Returns the queue position (1-based).
func (qm *QueueManager) Enqueue(job *dwsv1alpha1.DWSJob) int32 {
	qm.mu.Lock()
	defer qm.mu.Unlock()

	queueName := job.Spec.QueueName
	if queueName == "" {
		queueName = "default"
	}

	entry := queueEntry{
		name:      job.Name,
		namespace: job.Namespace,
		priority:  job.Spec.Priority,
	}

	// Check if already in queue.
	for i, e := range qm.queues[queueName] {
		if e.name == job.Name && e.namespace == job.Namespace {
			return int32(i + 1)
		}
	}

	qm.queues[queueName] = append(qm.queues[queueName], entry)

	// Sort by priority descending, then by name (stable proxy for creation time).
	sort.SliceStable(qm.queues[queueName], func(i, j int) bool {
		if qm.queues[queueName][i].priority != qm.queues[queueName][j].priority {
			return qm.queues[queueName][i].priority > qm.queues[queueName][j].priority
		}
		return qm.queues[queueName][i].name < qm.queues[queueName][j].name
	})

	// Return position.
	for i, e := range qm.queues[queueName] {
		if e.name == job.Name && e.namespace == job.Namespace {
			return int32(i + 1)
		}
	}

	return int32(len(qm.queues[queueName]))
}

// Dequeue removes a job from its queue.
func (qm *QueueManager) Dequeue(job *dwsv1alpha1.DWSJob) {
	qm.mu.Lock()
	defer qm.mu.Unlock()

	queueName := job.Spec.QueueName
	if queueName == "" {
		queueName = "default"
	}

	entries := qm.queues[queueName]
	for i, e := range entries {
		if e.name == job.Name && e.namespace == job.Namespace {
			qm.queues[queueName] = append(entries[:i], entries[i+1:]...)
			return
		}
	}
}

// GetPosition returns the current queue position of a job (1-based). Returns 0 if not queued.
func (qm *QueueManager) GetPosition(job *dwsv1alpha1.DWSJob) int32 {
	qm.mu.RLock()
	defer qm.mu.RUnlock()

	queueName := job.Spec.QueueName
	if queueName == "" {
		queueName = "default"
	}

	for i, e := range qm.queues[queueName] {
		if e.name == job.Name && e.namespace == job.Namespace {
			return int32(i + 1)
		}
	}
	return 0
}

// CanAdmit checks whether a job can be admitted from its queue based on
// the per-queue concurrency limit. It counts how many jobs in this queue
// are currently in Provisioning or Running phase.
func (qm *QueueManager) CanAdmit(ctx context.Context, job *dwsv1alpha1.DWSJob) (bool, error) {
	logger := log.FromContext(ctx)

	queueName := job.Spec.QueueName
	if queueName == "" {
		queueName = "default"
	}

	// Check if this job is at the front of its priority group.
	qm.mu.RLock()
	position := int32(0)
	for i, e := range qm.queues[queueName] {
		if e.name == job.Name && e.namespace == job.Namespace {
			position = int32(i + 1)
			break
		}
	}
	qm.mu.RUnlock()

	// Only the first job in the queue can be admitted.
	if position != 1 {
		logger.V(1).Info("Job not at front of queue", "position", position, "queue", queueName)
		return false, nil
	}

	// Count active jobs in this queue.
	var allJobs dwsv1alpha1.DWSJobList
	if err := qm.client.List(ctx, &allJobs); err != nil {
		return false, err
	}

	activeCount := 0
	for _, j := range allJobs.Items {
		jQueue := j.Spec.QueueName
		if jQueue == "" {
			jQueue = "default"
		}
		if jQueue != queueName {
			continue
		}
		if j.Status.Phase == dwsv1alpha1.DWSJobProvisioning || j.Status.Phase == dwsv1alpha1.DWSJobRunning {
			activeCount++
		}
	}

	if activeCount >= qm.maxConcurrentPerQueue {
		logger.V(1).Info("Queue at concurrency limit",
			"queue", queueName,
			"active", activeCount,
			"limit", qm.maxConcurrentPerQueue,
		)
		return false, nil
	}

	return true, nil
}

// RebuildQueues rebuilds the in-memory queue state from the cluster.
// This is called on controller startup to recover state.
func (qm *QueueManager) RebuildQueues(ctx context.Context) error {
	logger := log.FromContext(ctx)

	var allJobs dwsv1alpha1.DWSJobList
	if err := qm.client.List(ctx, &allJobs); err != nil {
		return err
	}

	qm.mu.Lock()
	defer qm.mu.Unlock()

	qm.queues = make(map[string][]queueEntry)

	for _, job := range allJobs.Items {
		if job.Status.Phase != dwsv1alpha1.DWSJobQueued && job.Status.Phase != dwsv1alpha1.DWSJobPending {
			continue
		}

		queueName := job.Spec.QueueName
		if queueName == "" {
			queueName = "default"
		}

		entry := queueEntry{
			name:      job.Name,
			namespace: job.Namespace,
			priority:  job.Spec.Priority,
		}
		qm.queues[queueName] = append(qm.queues[queueName], entry)
	}

	// Sort each queue.
	for queueName := range qm.queues {
		sort.SliceStable(qm.queues[queueName], func(i, j int) bool {
			if qm.queues[queueName][i].priority != qm.queues[queueName][j].priority {
				return qm.queues[queueName][i].priority > qm.queues[queueName][j].priority
			}
			return qm.queues[queueName][i].name < qm.queues[queueName][j].name
		})
		logger.Info("Rebuilt queue", "queue", queueName, "size", len(qm.queues[queueName]))
	}

	return nil
}
